src/FormControl/FormControl.tsx

```diff
@@ -0,0 +1,232 @@
+import React from 'react'
+import {Autocomplete, Box, Checkbox, Radio, Select, Textarea, TextInput, TextInputWithTokens, useSSRSafeId} from '..'
+import InputValidation from '../_InputValidation'
+import {ComponentProps} from '../utils/types'
+import FormControlCaption from './_FormControlCaption'
+import FormControlLabel from './_FormControlLabel'
+import FormControlValidation from './_FormControlValidation'
+import {Slots} from './slots'
+import ValidationAnimationContainer from '../_ValidationAnimationContainer'
+import {get} from '../constants'
+import FormControlLeadingVisual from './_FormControlLeadingVisual'
+
+export interface Props {
+  children?: React.ReactNode
+  /**
+   * Whether the field is ready for user input
+   */
+  disabled?: boolean
+  /**
+   * The unique identifier for this field. Used to associate the label, validation text, and caption text
+   */
+  id?: string
+  /**
+   * Whether this field must have a value for the user to complete their task
+   */
+  required?: boolean
+  /**
+   * Changes the layout of the form control
+   */
+  variant?: 'stack' | 'choice' // TODO: come up with a better name for 'stack'
+}
+
+type FormControlValidationProps = ComponentProps<typeof FormControlValidation>
+export interface FormControlContext extends Pick<Props, 'disabled' | 'id' | 'required'> {
+  captionId: string
+  validationMessageId: string
+}
+
+const FormControl = ({children, disabled, id: idProp, required, variant = 'stack'}: Props) => {
+  const expectedInputComponents = [Autocomplete, Checkbox, Radio, Select, TextInput, TextInputWithTokens, Textarea]
+  const id = useSSRSafeId(idProp)
+  const validationChild = React.Children.toArray(children).find(child =>
+    React.isValidElement<FormControlValidationProps>(child) && child.type === FormControlValidation ? child : null
+  )
+  const captionChildren: React.ReactElement[] | undefined | null = React.Children.map(children, child =>
+    React.isValidElement(child) && child.type === FormControlCaption ? child : null
+  )?.filter(Boolean)
+  const labelChild: React.ReactNode | undefined | null = React.Children.toArray(children).find(
+    child => React.isValidElement(child) && child.type === FormControlLabel
+  )
+  const validationMessageId = validationChild ? `${id}-validationMsg` : ''
+  const validationStatus = React.isValidElement(validationChild) ? validationChild.props.appearance : undefined
+  const captionId = captionChildren?.length ? `${id}-caption` : undefined
+  const InputComponent = React.Children.toArray(children).find(child =>
+    expectedInputComponents.some(inputComponent => React.isValidElement(child) && child.type === inputComponent)
+  )
```

Comment:
Does this implementation prevent people from passing custom input components?

Reply:
Yes. We can open it up later if we need to.